### PR TITLE
feat(root_finding): add Bernstein polynomial root-finding module

### PR DESCRIPTION
**Pull request by DavorDobrota**
## Summary

- Ports root-finding algorithms from `parametric_immersion` into pantr as a new `root_finding` subpackage
- Implements two complementary solvers: **Yuksel's monotone decomposition** (scale-invariant, low overhead for moderate degrees) and **Bezier clipping** (super-linear convergence for high-degree well-conditioned polynomials)
- Auto-dispatcher selects the best algorithm based on polynomial degree and coefficient dynamic range (thresholds tuned from a 27k-trial sweep)
- Batch-parallel variants (`find_roots_batch`, `solve_monotone_root_batch`) use `nb_prange` for multi-core throughput

### Architecture (pantr 3-layer pattern)

| Layer | Files | Role |
|-------|-------|------|
| L1 Public API | `_root_finding.py` | `find_roots`, `find_roots_batch`, `solve_monotone_root`, `solve_monotone_root_batch` |
| L2 Validation | `_find_roots.py` | Input normalization, tolerance resolution, dispatch |
| L3 Numba kernels | `_root_finding_core.py`, `_yuksel_core.py`, `_clipping_core.py`, `_batch_core.py` | All computation, `@nb_jit(nopython=True, cache=True)` |

## Test plan

- [x] 66 unit tests covering all layers, algorithms, edge cases, batch ops
- [x] 55 stress tests cross-validating Yuksel vs clipping on 2,050 randomized polynomials (random, wiggly, near-zero, extreme range, double roots, rational ray equations)
- [x] Batch consistency verified against single-polynomial results
- [x] `ruff check` clean
- [x] `ruff format` clean
- [x] `mypy --strict` clean
- [x] Full test suite passes (no regressions)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/pantr/root_finding/_batch_core.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            j -= 1
+        raw_roots[j + 1] = key
+
+    # Basic dedup within the kernel (no derivative-aware merge in Numba).
+    coeff_scale = 0.0
+    for i in range(n + 1):
+        av = abs(coeff[i])
+        coeff_scale = max(coeff_scale, av)
+    zero_tol = max(coeff_scale * (n + 1) * 4.0 * _DBL_EPSILON, geom_tol)
+    dedup_tol = max(param_tol * 2.0, zero_tol * 4.0)
```
Comment (DavorDobrota) [inline src/pantr/root_finding/_batch_core.py]: This does not preserve the semantics of the original code. How can this be fixed?

**Comment by pantolin** [comment]
## Code Review for @DavorDobrota 

### Bugs

**1. Empty-batch crash**

`_find_roots_batch_impl` and `_solve_monotone_root_batch_impl` both resolve tolerance via `arr[0]` without guarding for `n_polys == 0`. `_validate_coeffs_batch` checks `shape[1] >= 1` but not `shape[0] >= 1`, so:

```python
find_roots_batch(np.empty((0, 3), dtype=np.float64))  # IndexError: arr[0]
```

Fix: add `if n_polys == 0: return ...` before the `arr[0]` access, or guard in `_validate_coeffs_batch`.

---

### Behavioral Inconsistency

**2. Single vs. batch dedup differ**

`find_roots` (single path) uses `_dedup_roots` from `_clipping_core.py`, which has a derivative-aware adaptive merge radius (`local_tol = zero_tol / max(abs(df), _DBL_EPSILON)`). The batch path uses `_dispatch_and_find` in `_batch_core.py`, which has a simpler fixed-threshold dedup:

```python
dedup_tol = max(param_tol * 2.0, zero_tol * 4.0)  # no derivative-awareness
```

This means `find_roots` and `find_roots_batch` can return different root counts for near-duplicate roots. The `TestBatchConsistency` stress tests pass because the random polynomials don't exercise this edge case. Consider noting this limitation in the `find_roots_batch` docstring, or unifying the dedup logic.

---

### Maintainability

**3. Dispatch thresholds duplicated across layers**

`_CLIP_MIN_DEGREE = 6` and `_CLIP_COEFF_RANGE_LIMIT = 1e8` are defined independently in both `_find_roots.py` (L2) and `_batch_core.py` (L3). The PR notes these were tuned from a 27k-trial sweep, so they're non-obvious. They can silently diverge if one is updated. Since Numba `nopython=True` kernels can't import Python-level module constants at call time, the only option is to either:
- Keep them co-located in `_batch_core.py` (L3 source of truth) and reference them via a comment in L2, or
- Add a module-level note in both files cross-referencing each other.

---

### Docstring Inaccuracies

**4. Wrong return type in `_dispatch_and_find` docstring** (`_batch_core.py:127`):

```
Returns:
    tuple[npt.NDArray[np.float32 | np.float64], int]:  # wrong
```

The function always allocates `np.empty(..., dtype=np.float64)`, so the return is `npt.NDArray[np.float64]`. Same issue in `_split_scalar` (`_root_finding_core.py`).

**5. `_clipping_core.py` module docstring** says "all Numba-compiled" in its exports list but `_dedup_roots` is a plain Python function.

---

### Minor Observations

- `_warmup_numba_functions` in `_batch_core.py` comments that it "transitively compiles all L3 helpers", but the parent `__init__.py` also calls all four modules' warmup functions explicitly — making three of the four calls redundant. The comment in `_batch_core._warmup_numba_functions` is misleading.

- `test_extreme_range_falls_back_to_yuksel` only checks that `len(roots) == n` (same count as Yuksel), not that the roots are actually valid. Worth adding a residual check.

---

### Positives

- Clean 3-layer adherence with no validation in L3 and no Numba in L2
- Derivative chain built iteratively (avoids Numba recursion limit) — correct approach
- `_subdivide_scalar` always operates on the original coefficients to avoid error accumulation — good
- Convex-hull clipping uses Andrew's monotone chain (O(n), no sort needed) — efficient
- Test coverage is excellent: boundary roots, double roots, near-zero, extreme range, rational ray equations, and algorithm cross-validation

---

### Summary

The two actionable items before merge are the **empty-batch crash** (clear bug) and **documenting the single-vs-batch dedup inconsistency** (behavioral). The duplicated dispatch thresholds are a future maintenance risk worth addressing. Everything else is minor. The algorithmic implementation looks sound.

**Comment by DavorDobrota** [comment]
Thanks for the thorough review! All points addressed in 79dc2e3:

1. **Empty-batch crash** — Fixed with early returns in both `_find_roots_batch_impl` and `_solve_monotone_root_batch_impl`. Added tests.
2. **Single vs batch dedup** — Documented in `find_roots_batch` docstring. Unifying would require making `_dedup_roots` Numba-compatible (it uses Python lists); not worth it for now.
3. **Duplicated thresholds** — Added cross-reference notes in both files. Numba nopython prevents sharing the constants via import.
4. **Return-type docstrings** — Fixed across all affected functions.
5. **Module docstring** — Split into "(Numba-compiled)" and "(plain Python)" labels.
6. **Warmup docstring** — Removed the misleading "transitively compiles all" claim.
7. **Test residual check** — `test_extreme_range_falls_back_to_yuksel` now validates residuals, not just count.

**Comment by pantolin** [comment]
@DavorDobrota Two refactoring commits have been pushed to this branch:

### 1. `refactor(root_finding): reuse bezier de Casteljau kernels`

- **Removed** `_de_casteljau_eval_scalar` and `_split_scalar` from `_root_finding_core.py` — these duplicated existing Bézier de Casteljau logic.
- **Added** two non-parallel scalar helpers to `_bezier_core.py`:
  - `_de_casteljau_eval_scalar` — single-parameter Bernstein evaluation via de Casteljau.
  - `_restrict_scalar` — Bernstein coefficient restriction to a sub-interval using the numerically stable two-pass strategy (pass ordering chosen to avoid dividing by a small number).
- **Refactored** `_slice_bezier_1d_core` and `_restrict_bezier_1d_core` to delegate per-column work to these scalar helpers (parallelism via `prange` over columns is preserved).
- **Rewrote** `_subdivide_scalar` in root-finding to be a thin adapter around `_restrict_scalar`, gaining the numerically superior pass-ordering strategy.
- Root-finding now **imports** `_de_casteljau_eval_scalar` from `_bezier_core` instead of defining its own.

### 2. `refactor(bezier): move root-finding module into bezier package`

- **Moved** all root-finding source files (`_batch_core.py`, `_clipping_core.py`, `_find_roots.py`, `_root_finding.py`, `_root_finding_core.py`, `_yuksel_core.py`) from `pantr.root_finding` to `pantr.bezier`.
- **Updated** `pantr.bezier.__init__` to export `find_roots`, `find_roots_batch`, `solve_monotone_root`, `solve_monotone_root_batch`.
- **Kept** `pantr.root_finding` as a backwards-compatible re-export shim.
- **Updated** all tests to import from `pantr.bezier`.

All root-finding and bezier tests pass.

**Comment by DavorDobrota** [comment]
@pantolin Do you think it is a good idea to re-export the root finding module? Since we have not merged it yet, maybe it is better to just deprecate it outright.

**Comment by pantolin** [comment]
@DavorDobrota 

a few comments about public interface:
- `solve_monotone_root_batch` should not return second argument (found) for consistency with `solve_monotone_root`. As said in the documentation, if no root is found, it returns a `NaN`.
- The input for `find_roots`, `find_roots_batch`, `solve_monotone_root` and `solve_monotone_root_batch` should be  a Bezier object (or list of Beziers).
  - Inside, it is needed to check that that Bezier dimension is equal to 1.
  - In the case of rational Beziers, only the first component is taken (x * w).
  - For batches, it requires to allocate an array of coefficients and copy all the coefs from all the Beziers. Maybe slow, but a pro-user can always access `_find_roots_batch_impl` or `_solve_monotone_root_batch_impl` directly.

**Comment by DavorDobrota** [comment]
@pantolin All the suggestions are applied.
